### PR TITLE
Deoptimization Engine

### PR DESCRIPTION
**Pull request by shyouhei**
### Abstract

Implemented a way to optimize ruby's executions and to revert them.  The strategy is restricted so that any VM states like program counter(s) would not be affected by the modifications.  This restriction makes deoptimize much lightweight.  Experiments show optimizations on this mechanism speeds up micro benchmarks, yet has no significant overheads for pitfall-ish activities, like eval.
### Introduction

<details>
<summary>

Ruby is said to be slow[[1](https://www.google.co.jp/search?q=ruby%20is%20slow)].  While there can be arguments on this, experiments show that it is at least not the fastest language to execute[[2](http://benchmarksgame.alioth.debian.org/u64q/which-programs-are-fastest.html)]. …
</summary>

The reason behind this is fingered as its GC being slow, or its GVL, or its dynamic nature.  But the author would like to point out the most important thing that Ruby lacks; it does not even try to optimize its execution.  We believe this is the root cause of the slowness we suffer.

We wrote "try to" above.  This is the essential concept of this pull request.  Why Ruby did not optimize so far was that its execution might be different on occasions.  For instance a `1 + 2` might not always be `3`, because the `+` can be redefined on-the-fly.  So the interpreter does not constant-fold the expression and looks for `Integer#+` every time.

This redefinition however, rarely happens.  Almost every time when `1 + 2` is evaluated, its value is arguably `3`.  Given this empirical assessment, it makes sense for Ruby to first try `3` and if `Integer#+` was redefined for any reason, fall back to that. It should be much faster for vast majority of normal (not redefined) cases, while it does not break redefinition with some overheads.

This sort of technique is called deoptimization[[3](http://dl.acm.org/citation.cfm?id=143114)].  Deoptimization is done more or less on other languages / implementations.  Notably JVM related projects has many research in this area. For instance JRuby+Truffle already has rich application of this technique[[4](http://chrisseaton.com/rubytruffle/deoptimizing/)].

What is pull-requested here is not the direct C translation of JRuby+Truffle's.  Let me describe what is going on.
### The deoptimization strategy

Since 1.9, Ruby's interpreter equips a language VM (visible as `::RubyVM` from inside ruby scripts).  It first generates sequences ("iseq"s hereafter) of VM instructions ("insn"s hereafter), and executes them in series.  The key structure of this mechanism is called `struct rb_iseq_constant_body`.

``` C
struct rb_iseq_constant_body {
    (…snip…)

    unsigned int iseq_size;
    const VALUE *iseq_encoded; /* encoded iseq (insn addr and operands) */

    (…snip…)
};
```

The `iseq_encoded` field is what VM scans to execute.  This is where we want to optimize.  Prior to do any modifications, we hold a backup of it using `memcpy`.  In case of deoptimization event,  we just overwrite this memory region with backed-up sequence using `memcpy` to revert whatever optimizations that might have been applied to it.

The upside of this strategy is that it is fairly simple.  Because everything is written in pure C code we have no portability concerns.  Also it is so simple that the deoptimization does not involve the VM states at all, including program counters.  This has advantage when execution of instructions recurs; if we had to update program counters, we must scan the whole VM stacks to keep consistencies between before / after deoptimization process.

The downside of this strategy is that it restricts what can be optimized.  For instance we cannot reorder instructions, because exceptions might raise inside of reordered sequences.  We cannot change the length of iseq, because the underlying memory region cannot be reallocated.

However, even with the given restriction, we can do something.
### Implemented optimizations

We have following optimizations in this pull request.
#### Constant folding

For instance a `1 + 2` is optimized into `3` like this.

``` diff
--- /dev/shm/xj79gt      2016-08-17 17:09:31.000000000 +0900
+++ /dev/shm/1gaaeo      2016-08-17 17:09:31.000000000 +0900
@@ -20,8 +20,10 @@ local table (size: 2, argc: 0 [opts: 0,
 |------------------------------------------------------------------------
 0000 trace            256                                             (   7)
 0002 trace            1                                               (   8)
-0004 putobject_OP_INT2FIX_O_1_C_
-0005 putobject        2
-0007 opt_plus         <callinfo!mid:+, argc:1, ARGS_SIMPLE>, <callcache>
+0004 putobject        3
+0006 nop
+0007 nop
+0008 nop
+0009 nop
 0010 trace            512                                             (   9)
 0012 leave                                                            (   8)
```

The `nop`s are introduced to meet "no PC modification" restriction.  In order to implement this, the `opt_plus` instruction is modified to detect a `putobject`-`putobject`-`opt-plus` sequence.  Then it calculates the computation as usual, and overwrites itself with the sequence like above diff.

Ruby-level constants (like `::Object`) are also subject to fold.  They are resolved and replaced to `putobject` with the identical value.
#### `send` elimination

Typical ruby script consists of many method invocations.  They are represented in `send` or its variant instructions in an iseq.  If they are optimized, the effect should not be marginal.  The problem is a method call is not always optimizable.  We have to determine whether we can skip calling.

In order to do so a method is called as usual for the first time, and checked whether it has any side effects or not.  If a method does not contain any side effects, it (more precisely, its underlying iseq) is marked as being "pure".  Next time when the same method is called, we can consider eliminating the call to it.

However a method that is not pure is not always NG to optimize.  A method can of course call another method.  A method that is not pure can be simply not checked yet.  So a method has 3 states namely "pure", "not pure", and "unpredictable".  Every methods start their life as being unpredictable.  Once its calling methods are all predicted to be pure, and itself has no side-effects, the state is updated to be pure.

``` C
enum insn_purity {
    insn_is_pure          = Qtrue,
    insn_is_not_pure      = Qfalse,
    insn_is_unpredictable = Qnil,
};
```

Side effect that we care includes:
- Writes to variables other than local ones,
- Definition of classes,
- Call of methods that are not pure,
- Call of C functions,
  - C functions are unable to check so we think they are all NG.
- Yielding blocks.

Once we know a method is pure, it is safe to eliminate a call to that method which immediately discards its return value(s).  Note however, that even on such situations, evaluations of method arguments are not always eligible to be eliminated.  They can have their own side effects.  We just eliminate the `send` variant instruction and not its caller site setup, like following diff:

``` diff
--- /dev/shm/165rrgd     2016-08-17 10:44:10.000000000 +0900
+++ /dev/shm/jd0rcj      2016-08-17 10:44:10.000000000 +0900
@@ -23,8 +23,10 @@ local table (size: 2, argc: 0 [opts: 0,
 0004 putself
 0005 putself
 0006 opt_send_without_block <callinfo!mid:n, argc:0, FCALL|VCALL|ARGS_SIMPLE>, <callcache>
-0009 opt_send_without_block <callinfo!mid:m, argc:1, FCALL|ARGS_SIMPLE>, <callcache>
-0012 adjuststack      1
+0009 adjuststack      2
+0011 nop
+0012 nop
+0013 nop
 0014 trace            1                                               (  16)
 0016 putnil
 0017 trace            512                                             (  17)
```

This is how a `m(n())` is optimized, where method `m` is pure but `n` is not.
#### Elimination of variable assignments

As we see, method calls can not be eliminated when its return value is used.  In order to relax this we would like to skip assignments of variables if possible.

However this is not easy to do precisely.  A variable's liveness shall be analyzed.  All types of variables except local ones are almost unable to analyze.  Also because we have restriction on what is allowed here due to deoptimization process, massive iseq rearrangements like SSA conversion is not doable.

In this pull request a very limited version of variable liveness analysis is implemented.  Only
local variables which are never referenced (assigned then made garbage immediately) are detected, then eliminated if possible.  This is not the only thing that could be done theoretically, but practically the easiest to detect so we do this as a bridgehead to further optimizations.

Note we have to check iseqs recursively; because local variables are reachable from inside of a block, and a block is represented in a separate iseq, which means we have to check all possible iseqs that can reach a local variable to ensure no references to it exists.

After checking for usages, a useless assignment can be eliminated like this:

``` diff
--- /dev/shm/1p1dusf     2016-08-25 18:06:58.000000000 +0900
+++ /dev/shm/1kpw0qe     2016-08-25 18:06:58.000000000 +0900
@@ -17,10 +17,14 @@ local table (size: 3, argc: 1 [opts: 0,
 [ 3] i<Arg>     [ 2] x
 0000 trace            256                                             (  12)
 0002 trace            1                                               (  13)
-0004 putself
-0005 getlocal_OP__WC__0 3
-0007 opt_send_without_block <callinfo!mid:m, argc:1, FCALL|ARGS_SIMPLE>, <callcache>
-0010 setlocal_OP__WC__0 2
+0004 nop
+0005 nop
+0006 nop
+0007 nop
+0008 nop
+0009 nop
+0010 nop
+0011 nop
 0012 trace            1                                               (  14)
 0014 putnil
 0015 trace            512                                             (  15)
```
### Experiments

To determine effectiveness of this approach we took experiments on my environment.  This machine equips `Intel(R) Core(TM)2 Duo CPU T7700` CPU and runs Ubuntu 16.04.  The image below is a `make benchmark` result against ruby 2.4.0dev.

Looking at the benchmark, the result is (generally speaking) similar to 2.4, with several exceptional cases which speeds up.  The most improved `vm2_bigarray*` got huge speedup because generation of a big array is eliminated at all.  Other improved benchmarks shows optimizations described above works quite well.

It is worth noting that the `vm2_eval*` result is 0.933 i.e. `eval` is only 2.7% slow compared to no optimization.  This can be considered as deoptimization preparation cost, and it means overhead is quite small.
### Conclusions

Implemented a deoptimization engine for ruby and several optimizations on top of it.  The proposed deoptimization strategy is lightweight so that activities like `eval` works without huge overheads.

The author is not willing to argue this is the only best optimization technique that we know.  Rather, even if it is suboptimal, the benchmark results show that the optimizations proposed here gains considerable amount of speed up.  This is why ruby is said to be slow; it is not optimized at all.  Any trivial optimizations should significantly impact it.

We believe ruby should have optimizations implemented.
### Future works

More optimizations can be thought on top of this deoptimization engine.  For instance branches with constant expressions could be relaxed to unconditional jumps.  More analysis on variable liveness could eliminate more of them.  Perversely, we could think of adding new local variables, like to eliminate common subexpressions.

The deoptimization strategy is subject to be improved.  If we can suffer more slowdown of `eval`, more fine-grained setup like three-address code could be thought of.  In order to do so a deoptimization engine should be able to revert IR conversion.

</details>

### CODE REVIEW

**Comment by headius** [comment]
I read through this a bit and have a question: can it deopt during a method body? If I'm reading it right, the deopt can only occur immediately before entry into the method, which means off-thread changes won't be seen until you re-enter the optimized method again. Is that correct?

**Comment by shyouhei** [comment]
@headius very good point.  My answer is yes and no.  On this patch deoptimizations can occur at two points:

- at the very beginning of a method, during construction of stack frame: https://github.com/ruby/ruby/pull/1419/commits/523120808bed6a54ced3cd8f17d1cf6a065678bf#diff-347392b62c9edab3c8ab5fb0d71a012fR196
    - This is expected to purge a stale optimized sequences.
- right after a method call returns. https://github.com/ruby/ruby/pull/1419/commits/523120808bed6a54ced3cd8f17d1cf6a065678bf#diff-d52fa5b396928e0544140fd48f329543R126
    - Redefinitions ultimately happen inside of some C functions defined elsewhere.  From a ruby script's point of view calling a C function is done via calling a method.  So every time we call a method, there are chances for redefinitions.  We have to make sure.

In short, inter-thread method tampering is not checked explicitly.  There are chances for deoptimizations during a method execution, but not immediately when another thread changed something.

I don't think we cannot know off-thread changes.  For now we have GVL so luckily we only need to check it right after when we acquire GVL.  That should suffice for now.  If we decide to give up GVL, we need to introduce some sort of thread checkpoint instead.

**Comment by shyouhei** [comment]
pushed https://github.com/ruby/ruby/pull/1419/commits/2f7bfbfdfc21f988525c6be276e8847384a5fbe9.  It checks for VM timestamp every time context switch hapens.

**Comment by chrisseaton** [comment]
Another option could be to have any operation which will cause deoptimisation to do the work to deoptimise the effected methods itself, before it transfers control back to them by releasing the GVL. Then the fast path (optimised methods) doesn't have to do a check. If they're still running, then they didn't have to deoptimise.

This is an advantage of the GVL. You shouldn't have to *ever* check if a method should deoptimise, if whoever causes the deoptimisation does the deoptimisation work itself. It only gets slightly more complicated with parallel threads, where you need to stop the world first.

**Comment by shyouhei** [comment]
@chrisseaton true.

However we now don't manage which method is optimized in which way so we have to start managing that part.  Because methods are subject to be GCed, management structure (if any) needs to have some sort of weak reference.  I think this is a bit too large to include in this pull request.

There is another possible way to know which one to deopt; scan all methods every time deoptimization is requested.  I doubt if this can be done in a timely fashion.

**Comment by headius** [comment]
@shyouhei Oh indeed, it should not be difficult to detect that there's been a deoptimization event across threads. The tricky bit will be deoptimizing the currently-executing iseqs without losing ipc, stack state, etc. I don't think that part exists yet, correct?

**Comment by shyouhei** [comment]
@headius no.  That (avoiding to touch VM states) is why we don't see considerable deopt overheads in above benchmarks.  @ko1 commented in-person that I would need to juggle VM states sooner or later I continue to develop, but not yet in this pull request.  It is still just patching iseqs.

**Comment by NullVoxPopuli** [comment]
Is there anything holding this back atm?

**Comment by jeremy** [comment]
Builds on `Apple LLVM version 8.0.0 (clang-800.0.40)` with a small change (else bus error on `memcpy`):
```diff
diff --git a/optimize.c b/optimize.c
index c0fee33..c9e666f 100644
--- a/optimize.c
+++ b/optimize.c
@@ -637,7 +637,7 @@ iseq_analyze(rb_iseq_t *iseq)
     }
 }
 
-static const VALUE wipeout_pattern[8]; /* maybe 5+2==7 should suffice? */
+static VALUE wipeout_pattern[8]; /* maybe 5+2==7 should suffice? */
 static VALUE adjuststack;
 static VALUE nop;
 static VALUE putobject;

```

**Comment by noahgibbs** [comment]
Is this moving forward? Is there anything we can do to help?

**Comment by jeremy** [comment]
> Is there anything we can do to help?

Some suggestions:
1. Try to run it with your own app or other software.
2. Try to break it, causing a crash or incorrect behavior.
3. Benchmark it.
4. Review and critique the idea and implementation.

**Comment by jamescook** [comment]
I'm seeing errors when running rake's test suite:
[rake_tests_output.txt](https://github.com/ruby/ruby/files/524533/rake_tests_output.txt)

and errors when running `bundle` for rails:
[rails_bundle_output.txt](https://github.com/ruby/ruby/files/524542/rails_bundle_output.txt)

**Comment by noahgibbs** [comment]
I get a different error on "make". I'm building with CLang on Mac.

[deopt_make_errors.txt](https://github.com/ruby/ruby/files/530447/deopt_make_errors.txt)

**Comment by shyouhei** [comment]
Thank you.  I'm looking at the logs.

**Comment by shyouhei** [comment]
OK, after close inspection I found several bugs and fixed them all.  It now bundles and runs Rails 5.0.0.1, at least for me.

**Comment by NullVoxPopuli** [comment]
how come the appveyor test results aren't public?

**Comment by noahgibbs** [comment]
Still doesn't compile on Mac with CLang. I have changes locally that allow it to build but segfault. I've attached the "git diff". I'll try to get it to build and run successfully.
[deopt_mac_patch.txt](https://github.com/ruby/ruby/files/554536/deopt_mac_patch.txt)

**Comment by shyouhei** [comment]
Hmm.  Thank you.  I've not tested this against clang so far.  Will take a look.

**Comment by rye** [comment]
Here is the Make output when I compile using `clang`.

```
	CC = clang
	LD = ld
	LDSHARED = clang -dynamic -bundle
	CFLAGS = -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wno-tautological-compare -Wno-parentheses-equality -Wno-constant-logical-operand -Wno-self-assign -Wunused-variable -Werror=implicit-int -Werror=pointer-arith -Werror=write-strings -Werror=declaration-after-statement -Werror=shorten-64-to-32 -Werror=implicit-function-declaration -Werror=division-by-zero -Werror=deprecated-declarations -Werror=extra-tokens   -pipe
	XCFLAGS = -D_FORTIFY_SOURCE=2 -fstack-protector -fno-strict-overflow -fvisibility=hidden -DRUBY_EXPORT -fPIE
	CPPFLAGS = -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -I. -I.ext/include/x86_64-darwin16 -I./include -I. -I./enc/unicode/data/8.0.0
	DLDFLAGS = -Wl,-undefined,dynamic_lookup -Wl,-multiply_defined,suppress -fstack-protector -Wl,-u,_objc_msgSend -Wl,-pie -framework CoreFoundation
	SOLIBS = -lgmp
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin16.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
translating probes probes.d
. ./vm_opts.h
generating id.h
id.h updated
generating opt_sc.inc
generating optinsn.inc
generating optunifs.inc
generating insns.inc
generating insns_info.inc
generating vmtc.inc
generating vm.inc
generating node_name.inc
generating known_errors.inc
known_errors.inc updated
generating vm_call_iseq_optimized.inc
./revision.h updated
/usr/local/opt/ruby/bin/ruby --disable=gems ./tool/gen_dummy_probes.rb ./probes.d > probes.dmyh
compiling main.c
compiling dmydln.c
compiling miniinit.c
compiling dmyext.c
generating miniprelude.c
miniprelude.c updated
compiling miniprelude.c
compiling array.c
compiling bignum.c
compiling class.c
compiling compar.c
compiling compile.c
compiling complex.c
compiling cont.c
compiling debug.c
compiling deoptimize.c
In file included from deoptimize.c:23:
./deoptimize.h:76:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
1 warning generated.
compiling dir.c
compiling dln_find.c
compiling encoding.c
compiling enum.c
compiling enumerator.c
compiling error.c
compiling eval.c
compiling file.c
compiling gc.c
compiling hash.c
compiling inits.c
compiling io.c
compiling iseq.c
In file included from iseq.c:25:
./deoptimize.h:76:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
1 warning generated.
compiling load.c
compiling marshal.c
compiling math.c
compiling node.c
compiling numeric.c
compiling object.c
compiling optimize.c
In file included from optimize.c:23:
./optimize.h:57:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
./optimize.h:86:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
./optimize.h:117:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
./optimize.h:129:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
./optimize.h:139:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
In file included from optimize.c:24:
./deoptimize.h:76:20: warning: unknown attribute 'leaf' ignored [-Wunknown-attributes]
    __attribute__((leaf));
                   ^
optimize.c:252:16: error: implicit conversion loses integer precision: 'const VALUE' (aka 'const unsigned long') to 'int'
      [-Werror,-Wshorten-64-to-32]
        return ptr[2];
        ~~~~~~ ^~~~~~
optimize.c:272:16: error: implicit conversion loses integer precision: 'const VALUE' (aka 'const unsigned long') to 'int'
      [-Werror,-Wshorten-64-to-32]
        return ptr[2];
        ~~~~~~ ^~~~~~
optimize.c:434:34: error: implicit conversion loses integer precision: 'const VALUE' (aka 'const unsigned long') to 'enum ruby_vminsn_type'
      [-Werror,-Wshorten-64-to-32]
    enum ruby_vminsn_type insn = argv[0];
                          ~~~~   ^~~~~~~
optimize.c:793:26: error: implicit conversion loses integer precision: 'unsigned long' to 'int' [-Werror,-Wshorten-64-to-32]
    int argc   = p->argc + m;
        ~~~~     ~~~~~~~~^~~
optimize.c:812:20: error: implicit conversion loses integer precision: 'long' to 'int' [-Werror,-Wshorten-64-to-32]
    int len    = n + m;
        ~~~      ~~^~~
6 warnings and 5 errors generated.
make: *** [optimize.o] Error 1
```

**Comment by claudiug** [comment]
any status with this one? Is almost one year old :)

**Comment by mensfeld** [comment]
Interested in this one as well. Any news?

**Comment by nijikon** [comment]
Me also.

**Comment by marysieek** [comment]
@pedroaliens @shyouhei any update on this issue?

**Comment by benlovell** [comment]
Folks, piling on doesn't help expedite an issue.

**Comment by shyouhei** [comment]
Hi, quick update is that it's still work in progress.  I got comments from other core-devs including @ko1 , so I think I need his review to merge this, once after I finished on peaceful update.

Thank you for your interest!

**Comment by NullVoxPopuli** [comment]
I successfully made and installed this, but when I tried running bundle install, I got a segfault:
```
$ sudo bundle install
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your
bundle as root will break this application for all non-root users on this machine.
Fetching gem metadata from https://rubygems.org/./usr/local/lib/ruby/2.4.0/net/http.rb:939: [BUG] Segmentation fault at 0x00000000000010
ruby 2.4.0dev (2016-07-22 deoptimization 55727) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0023 p:---- s:0138 e:000137 CFUNC  :wait_readable
c:0022 p:0701 s:0134 E:002290 METHOD /usr/local/lib/ruby/2.4.0/net/http.rb:939
c:0021 p:0008 s:0121 e:000120 METHOD /usr/local/lib/ruby/2.4.0/net/http.rb:869
c:0020 p:0076 s:0118 e:000117 METHOD /usr/local/lib/ruby/2.4.0/net/http.rb:864
c:0019 p:0045 s:0115 e:000114 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/net-http-persistent/lib/net/http/persistent.rb:702
c:0018 p:0530 s:0110 E:001a50 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/net-http-persistent/lib/net/http/persistent.rb:633
c:0017 p:0041 s:0097 e:000096 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/net-http-persistent/lib/net/http/persistent.rb:996
c:0016 p:0148 s:0084 e:000083 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/fetcher/downloader.rb:50
c:0015 p:0048 s:0075 e:000074 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/fetcher/downloader.rb:16
c:0014 p:0028 s:0067 e:000066 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/fetcher/compact_index.rb:116
c:0013 p:0161 s:0061 e:000060 BLOCK  /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/compact_index_client/updater.rb:50
c:0012 p:0048 s:0054 e:000053 METHOD /usr/local/lib/ruby/2.4.0/tmpdir.rb:89
c:0011 p:0026 s:0047 e:000046 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/compact_index_client/updater.rb:30
c:0010 p:0068 s:0040 e:000039 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/compact_index_client.rb:81
c:0009 p:0151 s:0035 e:000034 METHOD /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/compact_index_client.rb:97
```

**Comment by shyouhei** [comment]
This pull request was a bit too big for other core devs to review.  I changed my strategy to push things forward. As you see above several parts of this pull request were separated then proposed.  Some of them had already been merged, to gain ~1% speedup already.

Right now I am preparing another single-cut branch which brings some more boost.  Yet to be pull requested because it fails some tests right now.  I hope it can be rerouted soon.

**Comment by shyouhei** [comment]
Like I wrote above we are not going to merge this pull request as-is.  So close now.
